src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -298,6 +298,26 @@ public bool HasContentChanged(DocumentState oldState)
                 || oldState.textAndVersionSource != this.textAndVersionSource;
         }
 
+        /// <summary>
+        /// True if the Text has changed
+        /// </summary>
+        public bool HasTextChanged(DocumentState oldState)
+        {
+            if (oldState.sourceTextOpt != this.sourceTextOpt
+                || oldState.textAndVersionSource != this.textAndVersionSource)
+            {
+                return true;
+            }
+            else if (oldState._treeSource == this._treeSource)
+            {
+                return false;
+            }
+            else
+            {
+                return oldState._treeSource.GetValue().Version != this._treeSource.GetValue().Version;
+            }
```

Comment:
I am not sure why you are using treeSource at all here. I think this is all you need

oldState.sourceTextOpt != this.sourceTextOpt || oldState.textAndVersionSource != this.textAndVersionSource

this could miss a case where someone updating document text with exact same content. but that should be a rare case.

if you really want to detect such case, you should use http://source.roslyn.io/#Microsoft.CodeAnalysis/Text/SourceText.cs,976

Version will not give you any better precision than you using textAndVersinSource equality. Version is for people who can't access textAndVersionSource directly but want to have something cheap to see whether text has changed.

Reply:
fixed. thanks